### PR TITLE
Add gatk align and preprocess

### LINKED ISSUE #1212: new subworkflow: Gatk4/Align and Preprocess
This will add a subworkflow to follow the gatk4 best practices for aligning and preprocessing sequencing data to be used for the gatk variant calling workflows (joint germline, tumour normal, tumour paired, create panel of normals)

This will be written to accept either fastq files (what is usually used) or ubams (what gatk recommend) as input. 
- Ubams will be converted to fastq and aligned with BWA then gatk mergebamalignments is used to recover data from the unaligned reads to be used alongside their aligned counterparts.
- Fastq files are just aligned using BWA
-  gatk4 markduplicates is then used to mark duplicate reads and the files are then sorted and indexed
-  Finally the two step BQSR process is run to recalibrate files. gatk4 baserecalibrator and applybqsr
-  These reads can then be used for GATK variant calling

best practices: https://gatk.broadinstitute.org/hc/en-us/articles/360035535912-Data-pre-processing-for-variant-discovery

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by GCJMackenzie**
## PR checklist

Closes #1212 

This will add a subworkflow to follow the gatk4 best practices for aligning and preprocessing sequencing data to be used for the gatk variant calling workflows (joint germline, tumour normal, tumour paired, create panel of normals)

This will be written to accept either fastq files (what is usually used) or ubams (what gatk recommend) as input.

Ubams will be converted to fastq and aligned with BWA then gatk mergebamalignments is used to recover data from the unaligned reads to be used alongside their aligned counterparts.
Fastq files are just aligned using BWA
gatk4 markduplicates is then used to mark duplicate reads and the files are then sorted and indexed
Finally the two step BQSR process is run to recalibrate files. gatk4 baserecalibrator and applybqsr
These reads can then be used for GATK variant calling

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: subworkflows/nf-core/gatk_align_and_preprocess/main.nf [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+//
+include { BWAMEM2_INDEX } from '../../../modules/bwamem2/index/main.nf'
+include { SAMTOOLS_FASTQ } from '../../../modules/samtools/fastq/main.nf'
+include { BWAMEM2_MEM   } from '../../../modules/bwamem2/mem/main.nf'
+include { GATK4_MERGEBAMALIGNMENT } from '../../../modules/gatk4/mergebamalignment/main.nf'
+include { PICARD_MARKDUPLICATES } from '../../../modules/picard/markduplicates/main.nf'
+include { PICARD_SORTSAM } from '../../../modules/picard/sortsam/main.nf'
+include { SAMTOOLS_INDEX } from '../../../modules/samtools/index/main.nf'
+include { GATK4_BASERECALIBRATOR } from '../../../modules/gatk4/baserecalibrator/main.nf'
+include { GATK4_APPLYBQSR } from '../../../modules/gatk4/applybqsr/main.nf'
```
Comment (maxulysse) [inline subworkflows/nf-core/gatk_align_and_preprocess/main.nf:13]:
```suggestion
include { BWAMEM2_INDEX           } from '../../../modules/bwamem2/index/main.nf'
include { BWAMEM2_MEM             } from '../../../modules/bwamem2/mem/main.nf'
include { GATK4_APPLYBQSR         } from '../../../modules/gatk4/applybqsr/main.nf'
include { GATK4_BASERECALIBRATOR  } from '../../../modules/gatk4/baserecalibrator/main.nf'
include { GATK4_MERGEBAMALIGNMENT } from '../../../modules/gatk4/mergebamalignment/main.nf'
include { PICARD_MARKDUPLICATES   } from '../../../modules/picard/markduplicates/main.nf'
include { PICARD_SORTSAM          } from '../../../modules/picard/sortsam/main.nf'
include { SAMTOOLS_FASTQ          } from '../../../modules/samtools/fastq/main.nf'
include { SAMTOOLS_INDEX          } from '../../../modules/samtools/index/main.nf'
```

File: subworkflows/nf-core/gatk_align_and_preprocess/main.nf [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+workflow GATK_ALIGN_AND_PREPROCESS {
+    take:
+    input       // channel: [ val(meta), [ input ], [ input_index ], [] ]
+    fasta               // channel: /path/to/reference/fasta
+    fai                 // channel: /path/to/reference/fasta/index
+    dict                // channel: /path/to/reference/fasta/dictionary
+    bwaindex            // channel: name for panel of normals
```
Comment (maxulysse) [inline subworkflows/nf-core/gatk_align_and_preprocess/main.nf:21]:
```suggestion
    bwaindex            // channel: /path to bwa indexes
```

File: subworkflows/nf-core/gatk_align_and_preprocess/main.nf [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    //
+    //If no bwa index has been provided, use bwamem2 index to create one from the fasta file
+    //
+    if (bwaindex) {
+        ch_bwa_index = bwaindex
+    } else {
+        BWAMEM2_INDEX( fasta )
+        ch_bwa_index = BWAMEM2_INDEX.out.index
+        ch_versions = ch_versions.mix(BWAMEM2_INDEX.out.versions)
+    }
```
Comment (maxulysse) [inline subworkflows/nf-core/gatk_align_and_preprocess/main.nf:40]: I do think this should be done separately in a local `prepare_genome` subworkflow like in rnaseq

File: subworkflows/nf-core/gatk_align_and_preprocess/main.nf [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    bwa_mem_out         = BWAMEM2_MEM.out.bam.collect()                     // channel: [ val(meta), [ vcf ] ]
+    // mutect2_index    = GATK4_MUTECT2.out.tbi.collect()                     // channel: [ val(meta), [ tbi ] ]
+    // mutect2_stats    = GATK4_MUTECT2.out.stats.collect()                   // channel: [ val(meta), [ stats ] ]
+
+    // genomicsdb       = GATK4_GENOMICSDBIMPORT.out.genomicsdb               // channel: [ val(meta), [ genomicsdb ] ]
+
+    // pon_vcf          = GATK4_CREATESOMATICPANELOFNORMALS.out.vcf           // channel: [ val(meta), [ vcf.gz ] ]
+    // pon_index        = GATK4_CREATESOMATICPANELOFNORMALS.out.tbi           // channel: [ val(meta), [ tbi ] ]
+
+    // versions         = ch_versions                                         // channel: [ versions.yml ]
```
Comment (maxulysse) [inline subworkflows/nf-core/gatk_align_and_preprocess/main.nf:120]:
```suggestion
    versions         = ch_versions                                         // channel: [ versions.yml ]
```

**Comment by maxulysse** [comment]
I'll stop reviewing now that I noticed it was a draft.
Looks promising, still some code left from the PON subworkflow

File: subworkflows/nf-core/gatk_align_and_preprocess/main.nf [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+include { BWAMEM2_INDEX           } from '../../../modules/bwamem2/index/main.nf'
+include { SAMTOOLS_FASTQ          } from '../../../modules/samtools/fastq/main.nf'
+include { BWAMEM2_MEM             } from '../../../modules/bwamem2/mem/main.nf'
+include { GATK4_MERGEBAMALIGNMENT } from '../../../modules/gatk4/mergebamalignment/main.nf'
+include { PICARD_MARKDUPLICATES   } from '../../../modules/picard/markduplicates/main.nf'
+include { SAMTOOLS_INDEX          } from '../../../modules/samtools/index/main.nf'
+include { GATK4_BASERECALIBRATOR  } from '../../../modules/gatk4/baserecalibrator/main.nf'
+include { GATK4_APPLYBQSR         } from '../../../modules/gatk4/applybqsr/main.nf'
+include { PICARD_SORTSAM as PICARD_SORTSAM_UNMAPPED         } from '../../../modules/picard/sortsam/main.nf'
+include { PICARD_SORTSAM as PICARD_SORTSAM_DUPLICATESMARKED } from '../../../modules/picard/sortsam/main.nf'
```
Comment (maxulysse) [inline subworkflows/nf-core/gatk_align_and_preprocess/main.nf:15]:
```suggestion
include { BWAMEM2_INDEX                                     } from '../../../modules/bwamem2/index/main.nf'
include { SAMTOOLS_FASTQ                                    } from '../../../modules/samtools/fastq/main.nf'
include { BWAMEM2_MEM                                       } from '../../../modules/bwamem2/mem/main.nf'
include { GATK4_MERGEBAMALIGNMENT                           } from '../../../modules/gatk4/mergebamalignment/main.nf'
include { PICARD_MARKDUPLICATES                             } from '../../../modules/picard/markduplicates/main.nf'
include { SAMTOOLS_INDEX                                    } from '../../../modules/samtools/index/main.nf'
include { GATK4_BASERECALIBRATOR                            } from '../../../modules/gatk4/baserecalibrator/main.nf'
include { GATK4_APPLYBQSR                                   } from '../../../modules/gatk4/applybqsr/main.nf'
include { PICARD_SORTSAM as PICARD_SORTSAM_UNMAPPED         } from '../../../modules/picard/sortsam/main.nf'
include { PICARD_SORTSAM as PICARD_SORTSAM_DUPLICATESMARKED } from '../../../modules/picard/sortsam/main.nf'
```

**Comment by alanhoyle** [comment]
I note that this is showing up on the Modules board, not the Subworkflows board?

**Comment by lescai** [comment]
I believe this is superseded by https://github.com/nf-core/modules/issues/2161 and others

**Comment by GCJMackenzie** [comment]
I agree that this is superceeded now

**Pull request closed without merging**